Add Java implementation for Observer pattern #35

AditiVatish opened the pull request.

AditiVatish committed:
Subject Interface

AditiVatish committed:
Add classs Concrete to implement subject

AditiVatish committed:
Override methods

AditiVatish committed:
Interface observer

AditiVatish committed:
Concrete Observer

AditiVatish committed:
Main class  with object

AditiVatish committed:
Overrride method

AditiVatish committed:
Documentation to understand the concept and code better

Sumit-tech-joshi commented:
@AditiVatish The documentation.md file provides a concise overview of the Observer Pattern implementation in Java, explaining the purpose and functionality of each component involved. It includes clear usage guidelines for integrating the pattern into a project.

The obsercer_pattern.java file contains the actual implementation of the Observer Pattern, defining the Subject, ConcreteSubject, Observer, and ConcreteObserver classes. The code is well-written, demonstrating how to use the pattern effectively in a Java application. Overall, both files offer a solid foundation for understanding and implementing the Observer Pattern in Java.


Just one change i will suggest, there is a typo error in filename of observer pattern java file, can you please make it observer_pattern.java other than this everything looks good to me.

Sumit-tech-joshi commented on obsercer_pattern.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.ArrayList;
```
can you please name this file as observer_pattern.java

AditiVatish replied:
@Sumit-tech-joshi Thank You for let me know.

Dhuds1 commented on obsercer_pattern.java:
```diff
@@ -0,0 +1,75 @@
+
+    public void setState(int state) {
+        this.state = state;
+        notifyObservers();
```
Although automatically calling the `notifyObservers` method is completely acceptable, depending on use cases it may be better to manually call it,

AditiVatish replied:
@Dhuds1 Thank you for your suggestion let me call it manually.

Dhuds1 commented on obsercer_pattern.java:
```diff
@@ -0,0 +1,75 @@
+class ConcreteObserver implements Observer {
+    private ConcreteSubject subject;
+
+    public ConcreteObserver(ConcreteSubject subject) {
```
I don't know what this does, could you please explain it to me? I would like to learn!

AditiVatish replied:
@Dhuds1 By this code I was trying to make connection between observer and the subject by using ConcreteObserver class and its constructor ConcreteSubject.  In the constructor, the observer registers with the object, allowing it to receive notifications when the state of the object changes. The object calls the update() method when the state changes, allowing the observer to react accordingly.

Dhuds1 commented on obsercer_pattern.java:
```diff
@@ -0,0 +1,75 @@
+public class Main {
+    public static void main(String[] args) {
+        ConcreteSubject subject = new ConcreteSubject();
+        ConcreteObserver observer1 = new ConcreteObserver(subject);
```
OH, I think I understand, you pass in the subject that you're wanting to observer on the observer creation?

if you wanted to have an observer observe multiple subjects, how would you go about that?

I was going to suggest having conditional statements in your subscriber classes to see if an observer is already subscribed or not to a subject, but that would most likely be redundant?

Dhuds1 reviewed (commented):
Nice code! I added a few comments;

AditiVatish committed:
Remove Typo

AditiVatish committed:
Calling manually  notifyObservers

BasranS commented:
Your code for the Observer pattern in Java is good, but, I have some suggestions consider giving clearer names to the classes ConcreteSubject and ConcreteObserver so it's easier to understand what they do.  Instead of automatically adding observers in the constructor, you could create a separate method for adding them. This would make your code more flexible.

ash-teach commented on documentation.md:
```diff
@@ -0,0 +1,64 @@
+#Observer Pattern Implementation in Java
```
The formatting for this README.md file is mostly incorrect. Please open the file in VS Code and observe the file via the Preview Pane to see the issues:
- Headings are misapplied
- Paragraph breaks are missing
- Proper in-line code formatting has not been used

Please fix these issues and update the PR.

ash-teach commented on documentation.md:
```diff
@@ -0,0 +1,64 @@
+  3.Implement the Observer interface in your observer classes.
+  4.Use the addObserver, removeObserver, and notifyObservers methods in your subject class to manage observers
+    and notify them of state changes.
+
```
I would like to see a section of references. What resources did you consult in writing the above?

ash-teach commented on documentation.md:
```diff
@@ -0,0 +1,64 @@
+ 5. Main Class
+    This class contains the main method and demonstrates how to use the Observer pattern.
+    It creates a ConcreteSubject object and two ConcreteObserver objects.
+    It changes the state of the subject and observes how the observers react to the state changes. 
```
Having a description of the code is a good idea. There may be cases where providing some **short** snippets from your code could help to clarify your description. But please **do not** reproduce all code here - only short snippets as necessary.

ash-teach commented on documentation.md:
```diff
@@ -0,0 +1,64 @@
+  The Observer interface declares the method that subjects call to notify observers of a state change.
+
+##ConcreteObserver
+  The ConcreteObserver class implements the Observer interface. It registers itself with a subject during instantiation and updates its state whenever it receives a notification from the subject.
```
This is not best practice. See comments below.

ash-teach commented on observer_pattern.java:
```diff
@@ -0,0 +1,77 @@
+
+// Concrete Observer
+class ConcreteObserver implements Observer {
+    private ConcreteSubject subject;
```
This creates an unnecessary dependency. You should have no direct reference to th subject in your observer implementation.

ash-teach commented on observer_pattern.java:
```diff
@@ -0,0 +1,77 @@
+    public static void main(String[] args) {
+        ConcreteSubject subject = new ConcreteSubject();
+        ConcreteObserver observer1 = new ConcreteObserver(subject);
+        ConcreteObserver observer2 = new ConcreteObserver(subject);
```
Subscription of observers to the subject should happen here, they should not occur inside the observer constructor. In other words, you should have something similar to the following here:

```java
subject.addObserver(observer1)
```

ash-teach commented:
Thanks for this work, @AditiVatish. There are some errors in the implementation that must be fixed before this can be merged. Your documentation also needs quite a bit of work - make sure to open it in the VSCode Preview Pane to see how it would render. I also require references to any resources used.

AditiVatish committed:
Update README.md

AditiVatish committed:
Update documentation.md

AditiVatish committed:
Update observer_pattern.java

AditiVatish committed:
Update documentation.md

Commit:
add missing parameter

change parameter type to String to indicate that data should be passed via update - not the entire subject.

Commit:
Merge branch 'main' into Java-ovservational-Pattern
